File: lintcode/golang/1495_leaf_similar_trees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root1: the first tree
 * @param root2: the second tree
 * @return: returns whether the leaf sequence is the same
 */
func leafSimilar (root1 *TreeNode, root2 *TreeNode) bool {
    // write your code here.
    l1 := make([]int, 0)
    l2 := make([]int, 0)
    
    var helper func (node *TreeNode, l *[]int)
    helper = func (node *TreeNode, l *[]int) {
        if node == nil {
            return
        }
        
        if node.Left == nil && node.Right == nil {
            *l = append(*l, node.Val)
        }
        if node.Left != nil {
            helper(node.Left, l)
        }
        if node.Right != nil {
            helper(node.Right, l)
        }
    }
    
    helper(root1, &l1)
    helper(root2, &l2)
    
    if len(l1) != len(l2) {
        return false
    }
    
    for i, v := range l1 {
        if v != l2[i] {
            return false
        }
    }
    
    return true
}

File: lintcode/python/1495_leaf_similar_trees.py
#!/usr/bin/python -t

# dfs and bst inorder

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root1: the first tree
    @param root2: the second tree
    @return: returns whether the leaf sequence is the same
    """
    def leafSimilar(self, root1, root2):
        # write your code here.
        leaves1 = []
        self.inorder(root1, leaves1)
        leaves2 = []
        self.inorder(root2, leaves2)
        return leaves1 == leaves2

    def inorder(self, node, leaves):
        if not node:
            return
        if not node.left and not node.right:
            return leaves.append(node.val)

        self.inorder(node.left, leaves)
        self.inorder(node.right, leaves)
